### DISCUSSION TITLE
Clarify: Article import: How to proceed with `convert_units`

### DISCUSSION THREAD

**Discussion by lentschi** [foodcoops/foodsoft#1226] [Decisions] [answered]
How should we proceed with the `convert_units` checkbox ("Keep current units, recompute unit quantity and price (like synchronize).") on CSV upload?
(Not really compatible with the new article system - Do we still need it?)

For reference see the [ported code in this fork](https://github.com/foodcoopsat/foodsoft_hackathon/blob/6a38f7e8cd15a925b814f15632fde62cee6e8397/app/models/article.rb#L202), the [code upstream](https://github.com/foodcoops/foodsoft/blob/b2deb0ed1f93b970a53685580802fed9e6c0e5c1/app/models/article.rb#L195) as well as the [open upstream issue](https://github.com/foodcoops/foodsoft/issues/1059) on that topic.

**Comment by lentschi** [comment]
I moved this from an issue to a discussion (It's just a question so far. But one we have to answer before the 5.0 release.)

**Comment by lentschi** [comment]
@bennibu, @wvengen  I now checked and this feature goes back as far as the [initial commit of this repo](https://github.com/foodcoops/foodsoft_hackathon/commit/5b9a7e05#diff-24503fd25ed68e6ebceee4951bc4f9b255b197278d4aa9d86ef9d5afe3f26beaR159) :sweat_smile:.  So, do you maybe have an idea about who is still using this and if we really need to maintain it?

(I don't know who first developed this and why, as I don't know where the original repo is.)

**Comment by wvengen** [comment]
I think I introduced the checkbox for article import (iirc). Initially, only the shared database would convert units, but then there were some foodcoops having suppliers that didn't use the shared database / were hard to integrate (each month the spreadsheet format would change slightly, requiring a tweak on the article import; this would allow foodcoops to make small changes to the spreadsheet before import - though I'm not sure anymore how exactly this was used).

Then there were also cases, where the user uploading / updating articles would just replace all articles, not caring about foodcoops units (often very small suppliers not delivering in bulk, e.g. honey, pre packaged cheese made on the farm, etc.). There the whole unit conversion would just make things more complicated.

**Reply by lentschi** [reply]
Okay, so but do you think it's still necessary to have this feature in foodsoft 5.0 or should I just remove it? 
(Is your foodcoop still using this or do you know of any other that does? Also, if you know of another developers who might know about this, please mention them here, so they can partake in the discussion.)

If someone still needs that, I'll gladly reimplement it on top of the extended unit system.
But I'd still need to know, _how_ to do that (i.e. what the exact expected behavior is).

Not knowing the use cases, I'd personally vote for removing that feature. (Users uploading csv files should now use UN/ECE format to avoid ambiguity or maintain their own foodsoft plugin doing quirky country specific conversions like this)

**Reply by lentschi** [reply]
Hm... anyone else has an answer to this? @foodcoops/contributors

**Reply by wvengen** [reply]
Without knowing it in detail, I would think that the unit system replaces the existing unit conversion. I think it _would_ be important to provide a clear explanation to foodcoops on how to move to the new unit system. In larger foodcoops, there may be a workgroup that can dive into this, but for smaller ones, I think it can be challenging to migrate (especially with many products from a wholesale supplier). A clear explanation would help. And, if possible, a way to do this without having to add each product manually again, would help.
Just my 2c :)

**Reply by RayOei** [reply]
> Hm... anyone else has an answer to this? @foodcoops/contributors

I have asked around and FCAdamnoord uses it. The person who handles the import massages the wholesaler excel with many regexp to get it to be importable. The sharredlist functionality doesn't work, as far as I understand, so the next question would be how the new unit/import mechanism would work. Notable this is using "BioRemeo" and I remember this to be a topic but not if it was (not/yet/longer) supported.

**Reply by RayOei** [reply]
Found it: #1174 😁 

The reworked CSV may contain different units (e.g. 250 g mushrooms or 170g Hummus) and as far as I understand it, those units are then imported and thus the option is used.
Assuming the new import just takes what it is given, it does seem the option has no use anymore.

**Reply by lentschi** [reply]
@wvengen 

> Without knowing it in detail, I would think that the unit system replaces the existing unit conversion.

Yes, I agree :+1: 

> In larger foodcoops, there may be a workgroup that can dive into this, but for smaller ones, I think it can be challenging to migrate

Well migrating shouldn't be too hard, since the actual conversion is done automatically (by the rails db migrations) and new unit fields can be populated (by the already implemented semi-auto-migration feature or manually) as foodcoops see fit.

However in the special case of the `convert_units` checkbox we'd have a breaking change if we remove it that needs to be explained to users, yes. (User's would have to re-format their input excels, if they still use inputs like "banana 10-12 St".)

> I think it would be important to provide a clear explanation to foodcoops on how to move to the new unit system. [...] A clear explanation would help.

Yes of course we'll do that. This will be implemented in #1106 and in some other issue that may not have been created yet, but we'll be doing that before releasing 5.0 of course!

**Reply by lentschi** [reply]
@RayOei 

> Notable this is using "BioRemeo" and I remember this to be a topic but not if it was (not/yet/longer) supported.

I'm not sure, if that's the case. I believe that this feature (the `convert_units` checkbox) is for some kind of different hard-coded format, that has nothing to do with BioRomeo:
If I read the source code correctly, the `convert_units` checkbox [supports ''KI" for "box" and "ST" for "piece" plus those supported by `ruby-units`](https://github.com/foodcoops/foodsoft_hackathon/blob/6a38f7e8cd15a925b814f15632fde62cee6e8397/app/models/article.rb#L202), whereas BioRomeo [handles all kinds of Dutch unit strings](https://github.com/foodcoops/foodsoft/pull/1173/files#diff-df6339890226907f6c544296fa1bfb0140ecd6d3945b64f1338ac053ab141789R17).

**Reply by wvengen** [reply]
I think we need to distinguish between
1. Import formats, like BioRomeo, which may recognize units and other things from the imported file
2. Unit conversion, which matches units in use by the foodcoop, to those in use by the supplier, and adjusts the foodcoop price

When `convert_units` is checked when importing a file (in whichever format), the foodcoop price is adjusted so that it matches the new price according to the units. 
The second is slated to be replaced by the new unit system (as far as I know).

**Reply by RayOei** [reply]
> @RayOei
> 
> > Notable this is using "BioRemeo" and I remember this to be a topic but not if it was (not/yet/longer) supported.
> 
> I'm not sure, if that's the case. I believe that this feature (the `convert_units` checkbox) 

@lentschi 

I don't suggest these are linked 😁 . FCANoord uses BioRemeo as their supplier. The information BR provide is reworked to the standard CSV format as there is no working BioRemeo import (it is outdated). The option is, currently, used as it apparently makes sure that the units used in that import CSV are taken as is. 
Therefore if the new import is able to just import what is in the CSV, like the 250 gr something and gr(ams) being an UNECE type, it seems the option is no longer needed (for FCANoord at least). We either need to adapt our current conversion if there is still a nonofficial type or we need to work on #1174 to have BR supported import.

**Reply by lentschi** [reply]
@wvengen 

> When convert_units is checked when importing a file (in whichever format), the foodcoop price is adjusted so that it matches the new price according to the units.
The second is slated to be replaced by the new unit system (as far as I know).

Well, I'm not _entirely_ sure, if it's _all_ replaced... The checkbox does multiple things at once, that aren't really well documented - as far as I understand the source code:

1. It converts the price from the source unit (article in the CSV) into the target unit (matching article in the DB) and leaves the target unit unchanged. If there is no matching article in the DB, it has _no effect_.
2. To determine _how to convert_ it allows specifying special units:
    a. When the source article is 'KI' (which translates to 'box') and the target unit is 'ST' ('piece'), it reads the amount fitting in a box from the _name_ using a regex.
    b. If a) is not applicable, it tries to parse both - the source article's unit as well as the target unit - using `ruby-units` and divides the price accordingly.

So, I believe, the article units part (2.) of all this is no longer needed: Instead of using "KI" for example, the UN/ECE equivalent "XBX" should be used.

The "price conversion and target unit remains unchanged" bit, however, might - in _theory_ - still be useful. It would require some effort to implement, but it could be done:

- The simple cases: For example, if a supplier used to list an article in kg but now lists it in gram, the price conversion logic already present since the article units PR could easily handle that.
- Where it gets more tricky: Consider a source article that comes in a box including pieces, each of which weighs 10 grams and a target article that came in pieces weighing 100 grams. This logic could still be implemented, but it would require a little more effort.
- Then of course there's cases, where articles just can't be converted, but those always existed and there's nothing that can be done about that.

But I wonder, should I really implement that?
I think the reason, this checkbox ever existed, was the limited way in which group orders worked before the article units PR: You couldn't order 0.4 boxes of an article. You _had_ to break it down into pieces on import. Now, however, you can just define the `group_order_unit` to be "piece", but still have the `supplier_order_unit` be "box".

So, in summary: I think this renders the whole checkbox obsolete - or do you see another use case why someone might need this? (There's a lot of assumptions on my end, so I might be wrong.)

**Reply by lentschi** [reply, marked as answer]
@foodcoops/contributors There seem to be no answers forthcoming. However, I feel we _need_ to start moving towards a release, so I'll choose the answer that makes the most sense to me:
Simply remove the feature -> I created an issue for that task: https://github.com/foodcoops/foodsoft/issues/1267

If someone still thinks we need it in there after all, we can always try to bring it back to live (see my thoughts on how to do that [above](https://github.com/foodcoops/foodsoft/discussions/1226#discussioncomment-15456141))
